crates/uv-redacted/src/lib.rs

```diff
@@ -1,21 +1,174 @@
-use std::borrow::Cow;
-
+use serde::{Deserialize, Serialize};
+use std::ops::{Deref, DerefMut};
+use std::str::FromStr;
 use url::Url;
 
-/// Return a version of the URL with redacted credentials, allowing the generic `git` username (without a password)
-/// in SSH URLs, as in, `ssh://git@github.com/...`.
-pub fn redacted_url(url: &Url) -> Cow<'_, Url> {
-    if url.username().is_empty() && url.password().is_none() {
-        return Cow::Borrowed(url);
+/// A URL wrapper that safely handles credentials for logging purposes.
+///
+/// `LogSafeUrl` wraps the standard [`url::Url`] type, providing functionality to mask
+/// secrets by default when the URL is displayed or logged. This helps prevent accidental
+/// exposure of sensitive information in logs and debug output.
+///
+/// # Examples
+///
+/// ```
+/// use uv_redacted::LogSafeUrl;
+/// use std::str::FromStr;
+///
+/// // Create a `LogSafeUrl` from a `&str`
+/// let mut url = LogSafeUrl::parse("https://user:password@example.com").unwrap();
+///
+/// // Display will mask secrets
+/// assert_eq!(url.to_string(), "https://user:****@example.com/");
+///
+/// // You can still access the username and password
+/// assert_eq!(url.username(), "user");
+/// assert_eq!(url.password(), Some("password"));
+///
+/// // And you can still update the username and password
+/// let _ = url.set_username("new_user");
+/// let _ = url.set_password(Some("new_password"));
+/// assert_eq!(url.username(), "new_user");
+/// assert_eq!(url.password(), Some("new_password"));
+///
+/// // It is also possible to remove the credentials entirely
+/// url.remove_credentials();
+/// assert_eq!(url.username(), "");
+/// assert_eq!(url.password(), None);
+/// ```
+#[derive(Clone, Eq, PartialEq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
+pub struct LogSafeUrl(Url);
```

Comment:
Maybe some more generic name, such as `RedactingUrl`?

Reply:
Yeah, I'm not sure what the best name is. I started with `ObfuscatedUrl` but worried that it missed the fact that the credentials are still _available_ (i.e., via `.username()` and `.password()`), just not displayed. 

At least `RedactingUrl` has the connotation that it actively does something, so that's good. But I'm worried it might still mislead developers that the credentials can't be accessed if needed. I don't have a strong opinion about it though.

Reply:
I'm not really sure what the best name is. I want to be careful not to imply that the credentials are in an obfuscated/redacted state internally, which is why I didn't go with `ObfuscatedUrl` initially. 

Your suggestion is better than that because it implies something active. But my worry would still be that it might mislead developers to think that the credentials can't be accessed at all (i.e., via `.username()` or `.password()`).

Reply:
I've gone with `DisplaySafeUrl` after some discussion. Updating PR name.